saltwater-repl/repl.rs

```diff
@@ -0,0 +1,226 @@
+        let config = Config::builder()
+            .history_ignore_space(true)
+            .history_ignore_dups(false)
+            .completion_type(CompletionType::List)
+            .edit_mode(EditMode::Emacs)
+            .max_history_size(1000)
+            .tab_stop(4)
+            .build();
+        let mut editor = Editor::with_config(config);
+
+        let commands = default_commands();
+        let help_message = generate_help_message(&commands);
+        let helper = ReplHelper::new(commands.iter().flat_map(|cmd| cmd.names).copied().collect());
+        editor.set_helper(Some(helper));
+
+        editor.bind_sequence(KeyPress::Up, Cmd::LineUpOrPreviousHistory(1));
+        editor.bind_sequence(KeyPress::Down, Cmd::LineDownOrNextHistory(1));
+        editor.bind_sequence(KeyPress::Tab, Cmd::Complete);
+
+        Self {
+            editor,
+            commands,
+            help_message,
+        }
+    }
+
+    pub fn run(&mut self) -> rustyline::Result<()> {
+        self.load_history();
+
+        println!("Saltwater {}", VERSION);
+        println!("Type {}help for more information.", PREFIX);
+        let result = loop {
+            let line = self.editor.readline(PROMPT);
+            match line {
+                Ok(line) => self.process_line(line),
+                // Ctrl + c will abort the current line.
+                Err(ReadlineError::Interrupted) => continue,
+                // Ctrl + d will exit the repl.
+                Err(ReadlineError::Eof) => break Ok(()),
+                Err(err) => break Err(err),
+            }
+        };
+        self.save_history();
+
+        result
+    }
+
+    pub fn save_history(&self) -> Option<()> {
+        let path = Self::history_path()?;
+        self.editor.save_history(&path).ok()
+    }
+
+    fn load_history(&mut self) -> Option<()> {
+        let path = Self::history_path()?;
+        self.editor.load_history(&path).ok()
+    }
+
+    fn history_path() -> Option<PathBuf> {
+        let mut history = data_dir()?;
+        history.push("saltwater_history");
+        Some(history)
+    }
+
+    fn process_line(&mut self, line: String) {
+        self.editor.add_history_entry(line.clone());
+
+        let trimmed_line = line.trim();
+        if trimmed_line.starts_with(PREFIX) {
+            let name = trimmed_line.split(' ').next().unwrap();
+
+            match self
+                .commands
+                .iter()
+                .find(|cmd| cmd.names.contains(&&name[1..]))
+            {
+                Some(cmd) => {
+                    let action = cmd.action;
+                    action(self, &trimmed_line[name.len()..])
+                }
+                None => println!("unknown command '{}'", name),
+            }
+        } else {
+            match self.execute_code(line) {
+                Ok(_) => {}
+                Err(err) => {
+                    // TODO: Proper error reporting
+                    println!("error: {}", err.data);
```

Comment:
This bit doesn't seem to be working (as mentioned above):
```
>> @
;
# no output
```

Reply:
because it's not a valid expression and thus get declined by rustyline.